Support work.result() to get result tensors for allreduce for Gloo, NCCL backends (resubmission of  #43386)

mrzzd opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43970 Support work.result() to get result tensors for allreduce for Gloo, NCCL backends (resubmission of  #43386)**

Original commit changeset: 27fbeb161706

Differential Revision: [D23455047](https://our.internmc.facebook.com/intern/diff/D23455047/)

Commit:
Back out "Revert D23216393: Support work.result() to get result tensors for allreduce for Gloo, NCCL backends"

Original commit changeset: 27fbeb161706

Differential Revision: [D23455047](https://our.internmc.facebook.com/intern/diff/D23455047/)

[ghstack-poisoned]

Commit:
Update on "Back out "Revert D23216393: Support work.result() to get result tensors for allreduce for Gloo, NCCL backends""


Original commit changeset: 27fbeb161706

Differential Revision: [D23455047](https://our.internmc.facebook.com/intern/diff/D23455047/)

[ghstack-poisoned]

pritamdamania87 commented:
We can close out this PR since the original diff has been reverted?

Commit:
Update on "Back out "Revert D23216393: Support work.result() to get result tensors for allreduce for Gloo, NCCL backends""


Original commit changeset: 27fbeb161706

Differential Revision: [D23455047](https://our.internmc.facebook.com/intern/diff/D23455047/)

[ghstack-poisoned]

pritamdamania87 commented on .circleci/config.yml:
```diff
@@ -6315,12 +6315,6 @@ workflows:
           name: pytorch_linux_xenial_cuda10_1_cudnn7_py3_multigpu_test
           requires:
             - pytorch_linux_xenial_cuda10_1_cudnn7_py3_gcc7_build
-          filters:
-            branches:
-              only:
-                - master
-                - /ci-all\/.*/
-                - /release\/.*/
```
What are the reason for these changes in the ci config?

mrzzd replied:
This is temporary change to force tests run for this PR. I'll undo it in next update to PR

pritamdamania87 commented on test/distributed/test_distributed.py:
```diff
@@ -857,6 +857,45 @@ def test_reduce_full_group_max(self):
+
+            result = work.result()
+            expected_value = 2 + (10 * (len(group) - 1))
+            self.assertEqual(result, [_build_tensor(src + 1, expected_value)])
```
nit: Since allreduce is in place we can just check if result is equal to tensor?

mrzzd replied:
> nit: Since allreduce is in place we can just check if result is equal to tensor?

Being equal to tensor does not mean result is ready, it could be both of them have non-finalized values. I see that's being tested in other unitttests, but still, better to not depend on them.

pritamdamania87 replied:
We are calling `work.wait()` before the check though right? That should synchronize the appropriate streams to ensure this is safe to do.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -810,6 +811,15 @@ class AsyncAllreduceWork : public ProcessGroupGloo::AsyncWork {
+
+
+  std::vector<at::Tensor> result() override {
+    TORCH_CHECK(isCompleted());
```
Add an error message to `TORCH_CHECK` here mentioning that work needs to be completed before calling `result()`.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -810,6 +811,15 @@ class AsyncAllreduceWork : public ProcessGroupGloo::AsyncWork {
+  std::vector<at::Tensor> outputs_;
 };
 
 class AsyncAllreduceCoalescedWork : public AsyncAllreduceWork {
```
Apart from `allreduce`, there are other ops like barrier, scatter, gather, reduce etc. can we add this functionality to those as well? Can be in a separate PR on top of this.

mrzzd replied:
Let's do it in separate PR

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -246,7 +246,8 @@ ProcessGroupNCCL::WorkNCCL::~WorkNCCL() {}
 bool ProcessGroupNCCL::WorkNCCL::isCompleted() {
   checkAndSetException();
-  return exception() || finishedGPUExecutionInternal();
+  bool result = exception() || finishedGPUExecutionInternal();
+  return result;
```
Why not have this in a single line as before?

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -782,6 +783,11 @@ std::shared_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
 }
 
+std::vector<at::Tensor> ProcessGroupNCCL::WorkNCCL::result() {
+  TORCH_CHECK(isCompleted());
```
This is not necessarily true for NCCL. `isCompleted()` returns true only when the work has completed successfully on the GPU. But we can still retrieve the `result` for the Work and perform async processing. I think we shouldn't have this check for NCCL.

mrzzd replied:
Good point, removed.

pritamdamania87 reviewed (changes requested):
If this is a resubmission of https://github.com/pytorch/pytorch/pull/43386, can we update the PR title and summary to reflect this appropriately?

Commit:
Update on "Back out "Revert D23216393: Support work.result() to get result tensors for allreduce for Gloo, NCCL backends""


Original commit changeset: 27fbeb161706

Differential Revision: [D23455047](https://our.internmc.facebook.com/intern/diff/D23455047/)

[ghstack-poisoned]

Commit:
Update on "Back out "Revert D23216393: Support work.result() to get result tensors for allreduce for Gloo, NCCL backends""


Original commit changeset: 27fbeb161706

Differential Revision: [D23455047](https://our.internmc.facebook.com/intern/diff/D23455047/)

[ghstack-poisoned]

Commit:
Update on "Back out "Revert D23216393: Support work.result() to get result tensors for allreduce for Gloo, NCCL backends""


Original commit changeset: 27fbeb161706

Differential Revision: [D23455047](https://our.internmc.facebook.com/intern/diff/D23455047/)

[ghstack-poisoned]

Commit:
Update on "Back out "Revert D23216393: Support work.result() to get result tensors for allreduce for Gloo, NCCL backends""


Original commit changeset: 27fbeb161706

Differential Revision: [D23455047](https://our.internmc.facebook.com/intern/diff/D23455047/)

[ghstack-poisoned]

mrzzd commented:
Verified that failing test also passed: https://app.circleci.com/pipelines/github/pytorch/pytorch/210229/workflows/86bde47b-f2da-48e3-a618-566ae2713102/jobs/7253683

Commit:
Update on "Back out "Revert D23216393: Support work.result() to get result tensors for allreduce for Gloo, NCCL backends""


Original commit changeset: 27fbeb161706

Differential Revision: [D23455047](https://our.internmc.facebook.com/intern/diff/D23455047/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -814,7 +814,7 @@ class AsyncAllreduceWork : public ProcessGroupGloo::AsyncWork {
 
   std::vector<at::Tensor> result() override {
-    TORCH_CHECK(isCompleted());
+    TORCH_CHECK(isCompleted(), "Work needs to be completed before calling result().");
```
nit: Extend this to mention, you should call `wait()` before `result()`.

pritamdamania87 reviewed (approved):
Looks good to me! Can we also update the PR summary appropriate? It currently still mentions "#43970 Back out "Revert D23216393: Support work.result() to get result tensors for allreduce for Gloo, NCCL backends"". Also, please check all the builds are clean before landing.

Commit:
Update on "Support work.result() to get result tensors for allreduce for Gloo, NCCL backends (resubmission of  #43386)"


Original commit changeset: 27fbeb161706

Differential Revision: [D23455047](https://our.internmc.facebook.com/intern/diff/D23455047/)

[ghstack-poisoned]

Commit:
Update on "Support work.result() to get result tensors for allreduce for Gloo, NCCL backends (resubmission of  #43386)"


Original commit changeset: 27fbeb161706

Differential Revision: [D23455047](https://our.internmc.facebook.com/intern/diff/D23455047/)

[ghstack-poisoned]

Commit:
Update on "Support work.result() to get result tensors for allreduce for Gloo, NCCL backends (resubmission of  #43386)"


Original commit changeset: 27fbeb161706

Differential Revision: [D23455047](https://our.internmc.facebook.com/intern/diff/D23455047/)

[ghstack-poisoned]

Commit:
Update on "Support work.result() to get result tensors for allreduce for Gloo, NCCL backends (resubmission of  #43386)"


Original commit changeset: 27fbeb161706

Differential Revision: [D23455047](https://our.internmc.facebook.com/intern/diff/D23455047/)

[ghstack-poisoned]

Commit:
Update on "Support work.result() to get result tensors for allreduce for Gloo, NCCL backends (resubmission of  #43386)"


Original commit changeset: 27fbeb161706

Differential Revision: [D23455047](https://our.internmc.facebook.com/intern/diff/D23455047/)

[ghstack-poisoned]

mrzzd commented:
> We can close out this PR since the original diff has been reverted?

I just rebased with latest changes in trunk. I am going to submit this as soon as testing finishes.
